### PR TITLE
sync docs with go reasons

### PR DESCRIPTION
**Pull request by ukff**
- add small go program which validates if declarated reasons are in sync with their metadata, if no it generates new md table which dev need to copy and insert into md file.
- update pipeline to execute it as test

### CODE REVIEW

**Review by IwonaLanger — APPROVED** [review]

**Review by IwonaLanger — DISMISSED** [review]

File: Makefile [hunk 1]
```diff
@@ -86,11 +86,15 @@ generate: controller-gen ## Generate code containing DeepCopy, DeepCopyInto, and
...
 .PHONY: test
-test: manifests kustomize generate fmt vet envtest ginkgo ## Run tests.
+test: manifests kustomize generate fmt vet envtest ginkgo  test-docs ## Run tests.
 	@. ./scripts/testing/set-env-vars.sh; \
 	go test -skip=TestAPIs ./... -timeout $(SUITE_TIMEOUT) -coverprofile cover.out -v; \
 	if [ "$(USE_EXISTING_CLUSTER)" == "true" ]; then $(GINKGO) -v controllers; else $(GINKGO) -v -p controllers; fi
 
+.PHONY: test-docs
+test-docs:
+	go run hack/autodoc/autodoc.go
```
Comment (jaroslaw-pieszka) [inline Makefile:96]: We need to move it to other location. `hack` is kind of reserved. We could move it to `scripts` or create new directory.

File: hack/autodoc/autodoc.go [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+		fmt.Println("Below can be found auto-generated table which contain new changes:")
+		fmt.Println(buildMdTable(reasonsMetadata))
+		os.Exit(errorExitCode)
+	}
+
+	os.Exit(okExitCode)
+}
+
+func extractData() string {
+	cmd := exec.Command("/bin/sh", "hack/autodoc/extract_conditions_data.sh")
```
Comment (jaroslaw-pieszka) [inline hack/autodoc/autodoc.go:67]: extract script name (you use it in main in message) to make it easier to change.

File: hack/autodoc/autodoc.go [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+	"fmt"
+	"os"
+	"os/exec"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+const (
+	spaceMargin             = 10
```
Comment (jaroslaw-pieszka) [inline hack/autodoc/autodoc.go:14]: Consider externalizing configuration (to allow adjusting without recompiling etc).

File: hack/autodoc/autodoc.go [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+				return true
+			}
+		}
+		return false
+	}
+	errors := make([]string, 0)
+
+	for _, constReason := range constReasons {
+		if !checkIfConstReasonHaveMetadata(constReason) {
+			errors = append(errors, fmt.Sprintf("there is a Reason = (%s) declarated in const scope, but there is no matching metadata for it", constReason))
```
Comment (jaroslaw-pieszka) [inline hack/autodoc/autodoc.go:124]:
```suggestion
			errors = append(errors, fmt.Sprintf("there is a Reason = (%s) declared in const scope, but there is no matching metadata for it", constReason))
```

File: hack/autodoc/autodoc.go [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+
+func mdTableToStruct(tableMd string) []reasonMetadata {
+	mdRows := strings.Split(tableMd, "\n")
+	mdRows = mdRows[2 : len(mdRows)-1]
+	structuredData := make([]reasonMetadata, 0)
+	for _, mdRow := range mdRows {
+		cleanLine := strings.Split(mdRow, "|")
+		cleanLine = cleanLine[1 : len(cleanLine)-1]
+
+		crState := cleanLine[1]
```
Comment (jaroslaw-pieszka) [inline hack/autodoc/autodoc.go:138]: Here (and below) we risk accessing non-existing elements (if MD table is corrupted).

File: hack/autodoc/autodoc.go [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+			errors = append(errors, fmt.Sprintf("Reason (%s) not found in docs.", newRow.conditionReason))
+		}
+	}
+	return errors
+}
+
+func buildMdTable(reasonsMetadata []reasonMetadata) string {
+	renderMdElement := func(length int, content string, spaceFiller string) string {
+		length = length - len(content)
+		element := ""
```
Comment (jaroslaw-pieszka) [inline hack/autodoc/autodoc.go:215]: Why? To let you use `+=` in the next line?

File: hack/autodoc/autodoc.go [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+	}
+
+	return mdTable.String()
+}
+
+func tryConvertGoLineToStruct(goLine string) (error, *reasonMetadata) {
+	if goLine == "" {
+		return fmt.Errorf("empty goLine given"), nil
+	}
+	goLine = strings.Replace(goLine, "\n", "", -1)
```
Comment (jaroslaw-pieszka) [inline hack/autodoc/autodoc.go:280]: Would strings.Trim suffice?

File: hack/autodoc/autodoc.go [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+	parts := strings.Split(goLine, "//")
+	if len(parts) != 2 {
+		return fmt.Errorf("in goLine (%s) there is no comment section (//) included, comment section should have following format (//CRState;Remark)", goLine), nil
+	}
+
+	words := strings.Fields(parts[0])
+	if len(words) != 2 {
+		return fmt.Errorf("goLine (%s) is bad structured, it should have following format (Reason: TypeAndStatus, //CRState;Remark", goLine), nil
```
Comment (jaroslaw-pieszka) [inline hack/autodoc/autodoc.go:288]:
```suggestion
		return fmt.Errorf("goLine (%s) is badly structured, it should have following format (Reason: TypeAndStatus, //CRState;Remark", goLine), nil
```

File: hack/autodoc/autodoc.go [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+	}
+
+	comments := strings.Split(parts[1], ";")
+	if len(comments) != 2 {
+		return fmt.Errorf("comment in goLine (%s) is bad structured, it should have following format (//CRState;Remark)", goLine), nil
```
Comment (jaroslaw-pieszka) [inline hack/autodoc/autodoc.go:293]:
```suggestion
		return fmt.Errorf("comment in goLine (%s) is badly structured, it should have following format (//CRState;Remark)", goLine), nil
```

**Review by jaroslaw-pieszka — CHANGES_REQUESTED** [review]

File: cmd/autodoc/main.go [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+	errors = compareContent(mdTableContent, reasonsMetadata)
+	if len(errors) > 0 {
+		printErrors(errors)
+		fmt.Println("Below can be found auto-generated table which contain new changes:")
+		fmt.Println(buildMdTable(reasonsMetadata))
+		os.Exit(errorExitCode)
+	}
+
+	os.Exit(okExitCode)
+}
```
Comment (MarekMichali) [inline cmd/autodoc/main.go:72]: In addition to the exit code, we could also add a log indicating that the script was executed correctly. Currently, if executed correctly, we get no information expect the code, and in the future someone who doesn't know the errorExitCode might wonder if it worked at all.

File: hack/autodoc/autodoc.go (hunk 3 above)
Comment (jaroslaw-pieszka) [reply hack/autodoc/autodoc.go]: After discussion we decided to postpone this task to next MVP.

**Review by jaroslaw-pieszka — APPROVED** [review]

**Review by IwonaLanger — APPROVED** [review]

**Pull request merged**